plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic.hs

```diff
@@ -349,9 +338,128 @@ tacticCmd _ _ _ _ =
        )
 
 
+------------------------------------------------------------------------------
+-- | Turn a 'RunTacticResults' into concrete edits to make in the source
+-- document.
+mkWorkspaceEdits
+  :: RunTacticResults
+  -> Judgement' a
+  -> RealSrcSpan
+  -> Context
+  -> DynFlags
+  -> LspFuncs c
+  -> Uri
+  -> Annotated ParsedSource
+  -> ( Either ResponseError Value
+     , Maybe (ServerMethod, ApplyWorkspaceEditParams)
+     )
+mkWorkspaceEdits rtr jdg span ctx dflags lf uri pm = do
+  let g = graftHole jdg (RealSrcSpan span) ctx rtr
+      response = transform dflags (clientCapabilities lf) uri g pm
+   in case response of
+        Right res -> (Right Null , Just (WorkspaceApplyEdit, ApplyWorkspaceEditParams res))
+        Left err -> (Left $ ResponseError InternalError (T.pack err) Nothing, Nothing)
+
+
+------------------------------------------------------------------------------
+-- | Graft a 'RunTacticResults' into the correct place in an AST. Correctly
+-- deals with top-level holes, in which we might need to fiddle with the
+-- 'Match's that bind variables.
+graftHole
+  :: Judgement' a2
+  -> SrcSpan
+  -> Context
+  -> RunTacticResults
+  -> Graft (Either String) ParsedSource
+graftHole jdg span ctx rtr
+  | _jIsTopHole jdg
+      = graftSmallestDeclsWithM span
+      $ graftDecl span
+      $ \pats ->
+        splitToDecl (fst $ last $ ctxDefiningFuncs ctx)
+      $ iterateSplit
+      $ mkFirstAgda (fmap unXPat pats)
+      $ unLoc
+      $ rtr_extract rtr
+graftHole jdg span _ rtr
+  = graftWithoutParentheses span
+    -- Parenthesize the extract iff we're not in a top level hole
+  $ bool maybeParensAST id (_jIsTopHole jdg)
+  $ rtr_extract rtr
+
+
+------------------------------------------------------------------------------
+-- | Merge in the 'Match'es of a 'FunBind' into a 'HsDecl'. Used to perform
+-- agda-style case splitting in which we need to separate one 'Match' into
+-- many, without affecting any matches which might exist but don't need to be
+-- split.
+mergeFunBindMatches
+    :: ([Pat GhcPs] -> LHsDecl GhcPs)
+    -> SrcSpan
+    -> HsBind GhcPs
+    -> Either String (HsBind GhcPs)
+mergeFunBindMatches make_decl span (fb@FunBind {fun_matches = mg@MG {mg_alts = L alts_src alts}}) =
```

Comment:
If you really want to do the split and re-parse, than doing the inverse of this would be the logical way to do it. for the declaration

```
foo True = 1
foo False = 2
```
shows up as two independent `Match` clauses, one for the `True` leg, one for the `False`.  So you can convert a given function declaration into one per match, process them individually, and then merge them. This mimics what already happens in GHC, where the relevant processing is in [cvtTopDecls](https://github.com/ghc/ghc/blob/22e13c1525c117919541437e60bc4352f0bc6517/compiler/parser/RdrHsSyn.hs#L339), and the actual work for a bind is at [getMonoBind](https://github.com/ghc/ghc/blob/22e13c1525c117919541437e60bc4352f0bc6517/compiler/parser/RdrHsSyn.hs#L391), which takes the first parsed declaration as the first param, and a list of the rest, and merges them,

Reply:
Good idea

Reply:
Much nicer, thanks for the idea.